bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java

```diff
@@ -266,6 +266,10 @@ public void run() {
         // We are about to send. Check if we need to make an ensemble change
         // because of delayed write errors
         lh.maybeHandleDelayedWriteBookieFailure();
+        // This op maybe recycled during bookie failures
+        if (maybeRecycled()) {
```

Comment:
So maybeHandleDelayedWriteBookieFailure triggers something that recycles this pending add op (I haven't looked that closely)?

Then the fix masks the problem. 
One side of the problem is that lh becomes null / results in NPE later. 
Another side of the problem is that pending add opp can be recycled and immediately reused by another add.

Reply:
> 

That's right. maybeHandleDelayedWriteBookieFailure will replace failed bookie with healthy one. However if NotEnoughBookieException is throwed. maybeHandleDelayedWriteBookieFailure will recycle this pending add op.
You can see this recycle logic in followed code logic
`

    @Override
    void handleBookieFailure(final Map<Integer, BookieId> failedBookies) {
        // handleBookieFailure should always run in the ordered executor thread for this
        // ledger, so this synchronized should be unnecessary, but putting it here now
        // just in case (can be removed when we validate threads)
        synchronized (metadataLock) {
            ...
            try {
                List<BookieId> newEnsemble = EnsembleUtils.replaceBookiesInEnsemble(
                        clientCtx.getBookieWatcher(), metadata, currentEnsemble, failedBookies, logContext);
            ...
            } catch (BKException.BKNotEnoughBookiesException e) {
                LOG.error("Could not get additional bookie to remake ensemble, closing ledger: {}", ledgerId);

                handleUnrecoverableErrorDuringAdd(e.getCode());
                return;
            }
`

Reply:
> So maybeHandleDelayedWriteBookieFailure triggers something that recycles this pending add op (I haven't looked that closely)?
> 
> Then the fix masks the problem. One side of the problem is that lh becomes null / results in NPE later. Another side of the problem is that pending add opp can be recycled and immediately reused by another add.

- One side of the problem is that lh becomes null / results in NPE later: 
If I understand correctly, only maybeHandleDelayedWriteBookieFailure has a chance to recycle this operation causing lh to become null which leads to this npe issue. Although this behavior is not as described in the followed comments about recycle condition, I think it is reasonable. 
```
    private void maybeRecycle() {
        /**
         * We have opportunity to recycle two objects here.
         * PendingAddOp#toSend and LedgerHandle#pendingAddOp
         *
         * A. LedgerHandle#pendingAddOp: This can be released after all 3 conditions are met.
         *    - After the callback is run
         *    - After safeRun finished by the executor
         *    - Write responses are returned from all bookies
         *      as BookieClient Holds a reference from the point the addEntry requests are sent.
         *
         * B. ByteBuf can be released after 2 of the conditions are met.
         *    - After the callback is run as we will not retry the write after callback
         *    - After safeRun finished by the executor
         * BookieClient takes and releases on this buffer immediately after sending the data.
         *
         * The object can only be recycled after the above conditions are met
         * otherwise we could end up recycling twice and all
         * joy that goes along with that.
         */
        if (hasRun && callbackTriggered) {
            ReferenceCountUtil.release(toSend);
            toSend = null;
        }
        // only recycle a pending add op after it has been run.
        if (hasRun && toSend == null && pendingWriteRequests == 0) {
            recyclePendAddOpObject();
        }
    }

```

- Another side of the problem is that pending add opp can be recycled and immediately reused by another add.: I don't understand this very well, can you explain in more detail?

Reply:
>> Another side of the problem is that pending add opp can be recycled and immediately reused by another add.

> I don't understand this very well, can you explain in more detail?

```java
        lh.maybeHandleDelayedWriteBookieFailure();
        
        // Add Op is recycled at this point, maybeRecycled() has a chance to check lh == null
        // but the recycler can give that op to another thread/request (via RECYCLER.get() call)
        // so it can be re-initialized with different lh and other data
        // so maybeRecycled returns false and the rest of the method is executed against different op
        // to avoid such problems and avoid locks we use ordered executors / guarantee order of calls
        
        if (maybeRecycled()) {
```

@zhaohaidao I hope this helps

Reply:
> > > Another side of the problem is that pending add opp can be recycled and immediately reused by another add.
> 
> > I don't understand this very well, can you explain in more detail?
> 
> ```java
>         lh.maybeHandleDelayedWriteBookieFailure();
>         
>         // Add Op is recycled at this point, maybeRecycled() has a chance to check lh == null
>         // but the recycler can give that op to another thread/request (via RECYCLER.get() call)
>         // so it can be re-initialized with different lh and other data
>         // so maybeRecycled returns false and the rest of the method is executed against different op
>         // to avoid such problems and avoid locks we use ordered executors / guarantee order of calls
>         
>         if (maybeRecycled()) {
> ```
> 
> @zhaohaidao I hope this helps

Thanks a lot for your explanation. I submitted a revision based on your suggestions.
Please have a look if you have time.
@dlg99